File: TTS.py
from key import service
from google.cloud import texttospeech
import os


def text_to_speech(title_output, text_input, gender_code):
    # 성별 지정
    if gender_code == 0:
        ssml_gender = texttospeech.SsmlVoiceGender.MALE
    else:
        ssml_gender = texttospeech.SsmlVoiceGender.FEMALE

    # 텍스트 입력
    text_input = texttospeech.SynthesisInput(ssml=text_input)

    # 음성 출력 설정
    voice = texttospeech.VoiceSelectionParams(
        language_code="ko-KR", ssml_gender=ssml_gender
    )

    # 음성 파일 형식 설정
    audio_config = texttospeech.AudioConfig(
        audio_encoding=texttospeech.AudioEncoding.MP3
    )

    # 목소리 설정
    response = client.synthesize_speech(
        input=text_input, voice=voice, audio_config=audio_config
    )

    # 음성 파일 작성

    # 음성 파일의 개수
    count_output = 0

    # 음성 파일 이름 = 'output' + '숫자' 의 숫자를 결정
    count_output = 0
    while True:
        if f'{title_output}_{count_output}.mp3' in files:
            count_output += 1
        else:
            break

    with open(f"{title_output}_{count_output}_py.mp3", "wb") as out:
        out.write(response.audio_content)
        print(f'Audio content written to file "{title_output}_{count_output}_py.mp3"')

# 탐색을 위한 로컬파일 경로 지정
target_dir = 'C:/Users/SSAFY/Desktop/TTS_test'
files = os.listdir(target_dir)

# 구글 인증 및 클라이언트 등록
service.connect()
client = texttospeech.TextToSpeechClient()

# 프로세스 실행
title_output = input('파일의 이름을 입력해주세요 : ')
text_input = input('텍스트를 입력해주세요 : ')
gender_code = int(input('성별을 입력해주세요 (남성:0, 여성:1) : '))

text_to_speech(title_output, text_input, gender_code)

File: TTS.js
const textToSpeech = require('@google-cloud/text-to-speech');
const fs = require('fs');
const util = require('util');
const service_js = require('./key/service_js.js');

// 임시 변수 설정
const title = 'test'
const text_input = 'test text'
const gender_input = 0

// TTS 함수
const TTS = (title, text_input, gender_input) => {
  // 파일 이름 겹치면 뒤에 숫자붙여서 저장
  let file_num = 0

  // 디렉토리 읽어오기
  fs.readdir('./', (err, fileList) => {
    
    //같은 이름 있으면 숫자 + 1
      for (filename of fileList) {
        if (`${title}_${file_num}_js.mp3` === filename) {
          file_num += 1
          }
        }
      }
  )

  // 성별 결정
  if (gender_input === 0) {
    gender = 'MALE'
    } else {
      gender = 'FEMALE'
    }

  // 서비스 연결
  service_js.connect()
  
  // 클라이언트 생성
  const client = new textToSpeech.TextToSpeechClient();

  // 실행 설정
  async function quickStart() {

    // text 읽어가기
    const text = text_input;

    // 설정
    const request = {
      // 텍스트로 받기
      input: {text: text},
      // 한국어, 성별
      voice: {languageCode: 'ko-KR', ssmlGender: gender},
      // mp3로 인코딩
      audioConfig: {audioEncoding: 'MP3'},
    };

    // 파일이름 설정 : writeFile 안에 있음
    const [response] = await client.synthesizeSpeech(request);
    const writeFile = util.promisify(fs.writeFile);
    await writeFile(`${title}_${file_num}_js.mp3`, response.audioContent, 'binary');
    console.log(`Audio content written to file: ${title}_${file_num}_js.mp3`);
  }
  quickStart();
}

// 함수 실행
TTS('test', 'test text', gender_input)